File: C++/greedy.cpp
// clang++ greedy.cpp -O3 -o greedy

#include <float.h>
#include <iostream>
#include <fstream>
#include <list>	
#include <boost/math/constants/constants.hpp>

using std::string;

class Activity 
{
public:
	string id;
	double lat;
	double lon;
};

class Resource 
{
public:
	string id;
	double lat;
	double lon;
};

class Allocation
{
public:
	string rid;
	string aid;
	double dist;
};

class SchemaData
{
public:
	std::list<Activity *> activity; 
	std::list<Resource *> resource; 
	std::list<Allocation *> allocation; 
};

const double pi = boost::math::constants::pi<double>();
const double earthRadius = 6367450.0; // geometric mean value gives about .1% error
const double convert2Rad = pi / 180.0;

double distanceBetweenPointsLatLong (double lat1, double lon1, double lat2, double lon2)
{
	double dStartLatInRad = lat1 * convert2Rad;
	double dStartLongInRad = lon1 * convert2Rad;
	double dEndLatInRad = lat2 * convert2Rad;
	double dEndLongInRad = lon2 * convert2Rad;
	double dLongitude = dEndLongInRad - dStartLongInRad;
	double dLatitude = dEndLatInRad - dStartLatInRad;
	double dSinHalfLatitude = sin(dLatitude * 0.5);
	double dSinHalfLongitude = sin(dLongitude * 0.5);
	double a = dSinHalfLatitude * dSinHalfLatitude + cos(dStartLatInRad) * cos(dEndLatInRad) * dSinHalfLongitude * dSinHalfLongitude;
	double c = atan2(sqrt(a), sqrt(1.0 - a));
	return (earthRadius * (c + c));
}

void importCSV(std::list<string>& lines, std::list<Activity *> *al, std::list<Resource *> *rl)
{
	string delimiter = ",";
	for (std::list<string>::iterator it = lines.begin(); it != lines.end(); ++it)
	{
		std::list<string> splits;
		size_t pos = 0;
		string token;
		while ((pos = (*it).find(delimiter)) != std::string::npos) {
		    token = (*it).substr(0, pos);
		    splits.push_back(token);
		    (*it).erase(0, pos + delimiter.length());
		}
		splits.push_back(*it);
		switch (splits.size())
		{
			case 3:
			{
				Resource *r = new Resource();
				r->id = splits.front();
				splits.pop_front();
				r->lat = atof(splits.front().c_str());
				splits.pop_front();
				r->lon = atof(splits.front().c_str());
				rl->push_back(r);
				break;
			}
			case 4:
			{
				Activity *a = new Activity();
				a->id = splits.front();
				splits.pop_front();
				a->lat = atof(splits.front().c_str());
				splits.pop_front();
				a->lon = atof(splits.front().c_str());
				al->push_back(a);
				break;
			}
		}
	}
}

void scheduleResources(SchemaData& sd) {
	for (std::list<Resource *>::iterator it = sd.resource.begin(); it != sd.resource.end(); ++it)
	{
		for (int c = 0; c < 50; c++) 
		{
			double lowest = DBL_MAX;
			std::list<Activity *>::iterator lowestobj;

			// See which is closest
			for (std::list<Activity *>::iterator ait = sd.activity.begin(); ait != sd.activity.end(); ++ait)
			{
				double dist = distanceBetweenPointsLatLong((*it)->lat, (*it)->lon, (*ait)->lat, (*ait)->lon);
				if (dist < lowest) 
				{
					lowest = dist;
					lowestobj = ait;
				}
			}
			Allocation *a = new Allocation();
			a->rid = (*it)->id;
			a->aid = (*lowestobj)->id;
			a->dist = lowest;
			sd.allocation.push_back(a);
			sd.activity.erase(lowestobj);
		}
	}
}

int main()
{
	// Read in CSV file
	std::list<string> csvLines;
	string line;
	std::ifstream csv ("/Users/daryl/Development/Projects/FunctionalComparison/Data/DataSPIF.csv");
	if (csv.is_open())
	{
		while (getline(csv, line))
		{
			csvLines.push_back(line);
		}
		csv.close();
	}
	else
	{
		std::cout << "Unable to open file" << std::endl << std::endl;
	}

	// Import each CSV line in turn
    std::list<Activity *> al;
	std::list<Resource *> rl;
	importCSV(csvLines, &al, &rl);

	for (int i = 0; i < 100	; i++)
	{
		SchemaData *sd = new SchemaData();
		sd->activity = std::list<Activity *>(al);
		sd->resource = std::list<Resource *>(rl);
		scheduleResources(*sd);	

		// Total
		double sum = 0.0;
		for (std::list<Allocation *>::iterator it = sd->allocation.begin(); it != sd->allocation.end(); ++it)
		{
			sum += (*it)->dist;
		}
		std::cout << i << " : " << sum << std::endl;

	}

	return 0;
}

File: Fortran/greedy.f90
module types

    type Activity
        character(50) :: id
        double precision :: lat
        double precision :: lon
        logical :: used 
    end type Activity

    type Resource
        character(50) :: id
        double precision :: lat
        double precision :: lon
    end type Resource

    type Allocation
        character(50) :: rid
        character(50) :: aid
        double precision :: dist
    end type Allocation

end module

module constants

    double precision, parameter :: pi = 3.141592653589793238462643
    double precision, parameter :: earthRadiusM = 6367450.0
    double precision, parameter :: convert2Rad = pi / 180.0
    double precision, parameter :: convert2Deg = 180.0 / pi

end module

pure function distance_between_points(lat1, lon1, lat2, lon2) result (distance)
    use constants
    implicit None

    double precision, intent (in) :: lat1
    double precision, intent (in) :: lon1
    double precision, intent (in) :: lat2
    double precision, intent (in) :: lon2
    double precision :: distance
    double precision :: dStartLatInRad, dStartLongInRad, dEndLatInRad, dEndLongInRad, dLongitude, dLatitude
    double precision :: dSinHalfLatitude, dSinHalfLongitude, a, b, c

    dStartLatInRad = lat1 * convert2Rad
    dStartLongInRad = lon1 * convert2Rad
    dEndLatInRad = lat2 * convert2Rad
    dEndLongInRad = lon2 * convert2Rad
    dLongitude = dEndLongInRad - dStartLongInRad
    dLatitude = dEndLatInRad - dStartLatInRad
    dSinHalfLatitude = sin(dLatitude * 0.5)
    dSinHalfLongitude = sin(dLongitude * 0.5)
    a = (dSinHalfLatitude * dSinHalfLatitude) + cos(dStartLatInRad) * cos(dEndLatInRad)
    b = a * (dSinHalfLongitude * dSinHalfLongitude)
    c = atan2(sqrt(b), sqrt(1.0 - b))
    distance = earthRadiusM * (c + c)
 end function

function schedule(activities, resources, allocations) result (distance)
    use types
    implicit None

    type(Activity), intent(inout) :: activities(500)
    type(Resource), intent(in) :: resources(10)
    type(Allocation), intent(inout) :: allocations(500)
    double precision :: distance

    integer :: i, j, l, k
    double precision :: this_distance
    double precision :: lowest_distance
    integer :: lowest_id
    double precision :: distance_between_points

    ! reset used
    activities%used = .false.

    ! Loop through each resource in turn
    k = 1
    do i = 1,10
        do l = 1,50
            lowest_distance = huge(0.0)
            lowest_id = -1
            do j = 1,500
                if (.not. activities(j)%used) then
                    this_distance = distance_between_points(resources(i)%lat, resources(i)%lon, &
                                                            activities(j)%lat, activities(j)%lon)
                    if (this_distance < lowest_distance) then
                        lowest_distance = this_distance
                        lowest_id = j
                    end if
                end if
            end do
            activities(lowest_id)%used = .true.
            allocations(k)%rid = resources(i)%id
            allocations(k)%aid = activities(lowest_id)%id
            allocations(k)%dist = lowest_distance
            k = k + 1
        end do
    end do

    ! add up distances
    distance = 0.0
    do i = 1,500
        distance = distance + allocations(i)%dist
    end do
end function

program greedy
    use types
    implicit None

    double precision :: result = 0
    character(50) :: id
    double precision :: lat, lon
    integer :: i  
    type(Activity) :: activities(500)
    type(Resource) :: resources(10)
    type(Allocation) :: allocations(500)
    character :: type
    double precision :: schedule

    ! Load input text file
    open(1, file = '../Data/DataSPIF.csv', status = 'old')
    do i = 1,510
        read (1,*) id, lat, lon
        type = id(1:1)
        if (type == 'R') then
            resources(i-500)%id = id
            resources(i-500)%lat = lat
            resources(i-500)%lon = lon
        else
            activities(i)%id = id
            activities(i)%lat = lat
            activities(i)%lon = lon
        end if
    end do
    close(1) 

    do i = 1,100
        result = schedule(activities, resources, allocations)
        print *, i, ': ', result
    end do

end program